### PR TITLE
SONAR-6591 Add SQALE metadata to XML rule format

### PR DESCRIPTION
**Pull request by simonbrandhof**
See javadoc of RulesDefinitionXmlLoader. It was deeply improved. 
Note that the branch is badly named. It's a 5.3 development.

### CODE REVIEW

File: sonar-plugin-api/src/main/java/org/sonar/api/server/rule/RulesDefinitionXmlLoader.java [hunk 1]
```diff
@@ -214,21 +303,38 @@ private void processRule(RulesDefinition.NewRepository repo, SMInputCursor ruleC
...
+          .setEffortToFixDescription(effortToFixDescription)
+          .setDebtSubCharacteristic(debtSubCharacteristic);
+      fillRemediationFunction(rule, debtRemediationFunction, debtRemediationFunctionOffset, debtRemediationFunctionCoeff);
+      fillParams(rule, params);
+    } catch (Exception e) {
+      throw new IllegalArgumentException(format("Fail to load the rule with key [%s:%s]", repo.key(), key), e);
+    }
+  }
+
+  private void fillRemediationFunction(RulesDefinition.NewRule rule, @Nullable String debtRemediationFunction, @Nullable String functionOffset, @Nullable String functionCoeff) {
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/server/rule/RulesDefinitionXmlLoader.java]: MAJOR Make "fillRemediationFunction" a "static" method. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2325)

File: sonar-plugin-api/src/main/java/org/sonar/api/server/rule/RulesDefinition.java [hunk 2]
```diff
@@ -613,11 +613,35 @@ public String toString() {
...
+    /**
+     * Shortcut for {@code create(Type.LINEAR, coefficient, null)}.
+     * @param coefficient TODO. See {@link DebtRemediationFunction} for details about format.
+     * @see org.sonar.api.server.debt.DebtRemediationFunction.Type#LINEAR
+     */
     DebtRemediationFunction linear(String coefficient);
 
+    /**
+     * Shortcut for {@code create(Type.LINEAR_OFFSET, coefficient, offset)}.
+     * @param coefficient TODO. See {@link DebtRemediationFunction} for details about format.
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/server/rule/RulesDefinition.java]: INFO Complete the task associated to this TODO comment. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1135)

File: sonar-plugin-api/src/main/java/org/sonar/api/server/rule/RulesDefinition.java [hunk 3]
```diff
@@ -613,11 +613,35 @@ public String toString() {
...
+     * @param offset TODO. See {@link DebtRemediationFunction} for details about format.
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/server/rule/RulesDefinition.java]: INFO Complete the task associated to this TODO comment. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1135)

File: sonar-plugin-api/src/main/java/org/sonar/api/server/rule/RulesDefinition.java [hunk 4]
```diff
@@ -613,11 +613,35 @@ public String toString() {
    * Factory of {@link org.sonar.api.server.debt.DebtRemediationFunction}.
    */
   interface DebtRemediationFunctions {
+
+    /**
+     * Shortcut for {@code create(Type.LINEAR, coefficient, null)}.
+     * @param coefficient TODO. See {@link DebtRemediationFunction} for details about format.
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/server/rule/RulesDefinition.java]: INFO Complete the task associated to this TODO comment. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1135)

**Comment by SonarTech** [comment]
SonarQube analysis reported 8 issues:
* MAJOR 5 major
* INFO 3 info

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR Add a 'package-info.java' file to document the 'protocol' package (org.sonarsource.sonarqube:sonar-batch-protocol:target/generated-sources/protobuf/org/sonar/batch/protocol) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1228)
* MAJOR Add a 'package-info.java' file to document the 'input' package (org.sonarsource.sonarqube:sonar-batch-protocol:target/generated-sources/protobuf/org/sonar/batch/protocol/input) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1228)
* MAJOR Add a 'package-info.java' file to document the 'output' package (org.sonarsource.sonarqube:sonar-batch-protocol:target/generated-sources/protobuf/org/sonar/batch/protocol/output) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1228)
* MAJOR [This method has 102 lines, which is greater than the 100 lines authorized. Split it into smaller methods.](https://github.com/SonarSource/sonarqube/blob/d426f4b6a195a9b4375a0dd4eba0cabcd97931d8/sonar-plugin-api/src/main/java/org/sonar/api/server/rule/RulesDefinitionXmlLoader.java#L222) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS138)

File: sonar-plugin-api/src/main/java/org/sonar/api/server/debt/DebtRemediationFunction.java [hunk 5]
```diff
@@ -23,27 +23,46 @@
...
+     * If you set the rule threshold to 20, and you have a paragraph with a complexity of 27, you have 7 points of complexity
+     * to remove. Internally, this is called the Effort to Fix. In that case, if you use the LINEAR_OFFSET configuration
+     * with an offset of 4h and a remediation cost of 1mn, the technical debt for this issue related to a
+     * too-complex block of code will be: (7 complexity points x 1min) + 4h = 4h and 7mn
+     * </p>
+     */
+    LINEAR_OFFSET(true, true),
+
+    /**
+     * The cost to fix all the issues of the rule is the same whatever the number of issues
```
Comment (julienlancelot) [inline sonar-plugin-api/src/main/java/org/sonar/api/server/debt/DebtRemediationFunction.java]: "whatever the number of issues" -> should be removed

Comment (simonbrandhof) [inline sonar-plugin-api/src/main/java/org/sonar/api/server/debt/DebtRemediationFunction.java]: Why ? That was extracted from doc of sonarsource.com.

File: sonar-plugin-api/src/main/java/org/sonar/api/server/rule/RulesDefinitionXmlLoader.java [hunk 6]
```diff
@@ -130,6 +191,11 @@ public void load(RulesDefinition.NewRepository repo, InputStream input, Charset
     }
   }
 
+  /**
+   * Loads rules by reading the XML input stream. The reader is closed by the method, so it
```
Comment (julienlancelot) [inline sonar-plugin-api/src/main/java/org/sonar/api/server/rule/RulesDefinitionXmlLoader.java]: is "not" closed

File: sonar-plugin-api/src/main/java/org/sonar/api/server/rule/RulesDefinitionXmlLoader.java [hunk 7]
```diff
@@ -202,10 +273,28 @@ private void processRule(RulesDefinition.NewRepository repo, SMInputCursor ruleC
         severity = StringUtils.trim(cursor.collectDescendantText(false));
 
       } else if (StringUtils.equalsIgnoreCase("cardinality", nodeName)) {
```
Comment (julienlancelot) [inline sonar-plugin-api/src/main/java/org/sonar/api/server/rule/RulesDefinitionXmlLoader.java]: StringUtils could be statically imported to improve readability

File: sonar-plugin-api/src/main/java/org/sonar/api/server/rule/RulesDefinitionXmlLoader.java [hunk 8]
```diff
@@ -202,10 +273,28 @@ private void processRule(RulesDefinition.NewRepository repo, SMInputCursor ruleC
...
-        cardinality = Cardinality.valueOf(StringUtils.trim(cursor.collectDescendantText(false)));
+        template = Cardinality.MULTIPLE == Cardinality.valueOf(StringUtils.trim(cursor.collectDescendantText(false)));
```
Comment (julienlancelot) [inline sonar-plugin-api/src/main/java/org/sonar/api/server/rule/RulesDefinitionXmlLoader.java]: All the if in this method are uglys, one day we should find a way to improve this...

File: sonar-plugin-api/src/main/java/org/sonar/api/server/debt/DebtRemediationFunction.java (hunk 5 above)
Comment (julienlancelot) [inline sonar-plugin-api/src/main/java/org/sonar/api/server/debt/DebtRemediationFunction.java]: Hum, other function was never depending of the number of issues, but I understand the fact to say that constant_issue depends on nothing...

File: sonar-plugin-api/src/main/java/org/sonar/api/server/rule/RulesDefinitionXmlLoader.java [hunk 9]
```diff
@@ -91,30 +103,79 @@
...
+ * &lt;rules&gt;
+ *   &lt;rule&gt;
+ *     &lt;key&gt;S1442&lt;/key&gt;
+ *     &lt;name&gt;"alert(...)" should not be used&lt;/name&gt;
+ *     &lt;description&gt;alert(...) can be useful for debugging during development, but ...&lt;/description&gt;
+ *     &lt;tag&gt;cwe&lt;/tag&gt;
+ *     &lt;tag&gt;security&lt;/tag&gt;
+ *     &lt;tag&gt;user-experience&lt;/tag&gt;
+ *     &lt;debtSubCharacteristic&gt;SECURITY_FEATURES&lt;/debtSubCharacteristic&gt;
+ *     &lt;debtRemediationFunction&gt;CONSTANT&lt;/debtRemediationFunction&gt;
```
Comment (julienlancelot) [inline sonar-plugin-api/src/main/java/org/sonar/api/server/rule/RulesDefinitionXmlLoader.java]: CONSTANT -> CONSTANT_ISSUE

**Comment by julienlancelot** [comment]
When using a unknown characteristic key, debt definition on the rule is simply ignored.
Is it expected ?

File: sonar-plugin-api/src/test/java/org/sonar/api/server/rule/RulesDefinitionXmlLoaderTest.java [hunk 10]
```diff
@@ -130,4 +142,114 @@ public void support_deprecated_format() {
...
+        "    <name>One</name>" +
+        "    <description>Desc</description>" +
+        "    <debtSubCharacteristic>BUG</debtSubCharacteristic>" +
+        "    <debtRemediationFunction>UNKNOWN</debtRemediationFunction>" +
+        "  </rule>" +
+        "</rules>");
+      fail();
+    } catch (IllegalArgumentException e) {
+      assertThat(e).hasMessageContaining("Fail to load the rule with key [squid:1]");
+      assertThat(e.getCause()).hasMessageContaining("No enum constant org.sonar.api.server.debt.DebtRemediationFunction.Type.UNKNOWN");
```
Comment (julienlancelot) [inline sonar-plugin-api/src/test/java/org/sonar/api/server/rule/RulesDefinitionXmlLoaderTest.java]: When trying to reproduce this use case, the error message doesn't contain "Fail to load the rule with key [squid:1]", only "No enum constant org.sonar.api.server.debt.DebtRemediationFunction.Type.UNKNOWN")

Comment (julienlancelot) [inline sonar-plugin-api/src/test/java/org/sonar/api/server/rule/RulesDefinitionXmlLoaderTest.java]: Begin of stack trace : 
2015.11.24 11:02:03 ERROR web[o.a.c.c.C.[.[.[/]] Exception sending context initialized event to listener instance of class org.sonar.server.platform.PlatformServletContextListener
java.lang.IllegalArgumentException: No enum constant org.sonar.api.server.debt.DebtRemediationFunction.Type.CONSTANT_ISSUE2
	at java.lang.Enum.valueOf(Enum.java:238) ~[na:1.8.0_65]
	at org.sonar.api.server.debt.DebtRemediationFunction$Type.valueOf(DebtRemediationFunction.java:34) ~[sonar-plugin-api-5.3-SNAPSHOT.jar:na]
	at org.sonar.api.server.rule.RulesDefinitionXmlLoader.fillRemediationFunction(RulesDefinitionXmlLoader.java:327) ~[sonar-plugin-api-5.3-SNAPSHOT.jar:na]
	at org.sonar.api.server.rule.RulesDefinitionXmlLoader.processRule(RulesDefinitionXmlLoader.java:318) ~[sonar-plugin-api-5.3-SNAPSHOT.jar:na]
	at org.sonar.api.server.rule.RulesDefinitionXmlLoader.load(RulesDefinitionXmlLoader.java:214) ~[sonar-plugin-api-5.3-SNAPSHOT.jar:na]

File: sonar-plugin-api/src/main/java/org/sonar/api/server/rule/RulesDefinitionXmlLoader.java [hunk 11]
```diff
@@ -159,76 +227,117 @@ private void processRule(RulesDefinition.NewRepository repo, SMInputCursor ruleC
...
+      } else if (equalsIgnoreCase("internalKey", nodeName)) {
+        internalKey = trim(cursor.collectDescendantText(false));
 
-      } else if (StringUtils.equalsIgnoreCase("priority", nodeName)) {
+      } else if (equalsIgnoreCase("priority", nodeName)) {
         // deprecated field, replaced by severity
-        severity = StringUtils.trim(cursor.collectDescendantText(false));
+        severity = trim(cursor.collectDescendantText(false));
+
+      } else if (equalsIgnoreCase("severity", nodeName)) {
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/server/rule/RulesDefinitionXmlLoader.java]: MAJOR This branch's code block is the same as the block for the branch on line 270. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1871)

File: sonar-plugin-api/src/main/java/org/sonar/api/server/rule/RulesDefinitionXmlLoader.java [hunk 12]
```diff
@@ -159,76 +227,117 @@ private void processRule(RulesDefinition.NewRepository repo, SMInputCursor ruleC
...
+          .setEffortToFixDescription(effortToFixDescription)
+          .setDebtSubCharacteristic(debtSubCharacteristic);
+      fillRemediationFunction(rule, debtRemediationFunction, debtRemediationFunctionOffset, debtRemediationFunctionCoeff);
+      fillParams(rule, params);
+    } catch (Exception e) {
+      throw new IllegalArgumentException(format("Fail to load the rule with key [%s:%s]", repo.key(), key), e);
+    }
+  }
+
+  private void fillRemediationFunction(RulesDefinition.NewRule rule, @Nullable String debtRemediationFunction,
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/server/rule/RulesDefinitionXmlLoader.java]: MAJOR Make "fillRemediationFunction" a "static" method. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2325)

**Comment by SonarTech** [comment]
SonarQube analysis reported 6 issues:
* MAJOR 6 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR Add a 'package-info.java' file to document the 'protocol' package (org.sonarsource.sonarqube:sonar-batch-protocol:target/generated-sources/protobuf/org/sonar/batch/protocol) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1228)
* MAJOR Add a 'package-info.java' file to document the 'input' package (org.sonarsource.sonarqube:sonar-batch-protocol:target/generated-sources/protobuf/org/sonar/batch/protocol/input) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1228)
* MAJOR Add a 'package-info.java' file to document the 'output' package (org.sonarsource.sonarqube:sonar-batch-protocol:target/generated-sources/protobuf/org/sonar/batch/protocol/output) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1228)
* MAJOR [This method has 102 lines, which is greater than the 100 lines authorized. Split it into smaller methods.](https://github.com/SonarSource/sonarqube/blob/eb04ce82a98b5b552d07cc7ddbb8dfc3886168b3/sonar-plugin-api/src/main/java/org/sonar/api/server/rule/RulesDefinitionXmlLoader.java#L224) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS138)

**Comment by julienlancelot** [comment]
@simonbrandhof Please fix some issues I've reported.

**Comment by simonbrandhof** [comment]
The sub-characteristic is ignored if it does not exist at runtime in the SQALE model. That's the expected behavior to support ability to change model with SonarSource SQALE plugin.

**Comment by julienlancelot** [comment]
ok

**Comment by simonbrandhof** [comment]
For the exception trap, the problem comes from picocontainer which does some trick (our code or pico ? I don't know) when container fails to stop. It's not straightforward to fix and will be done outside this ticket.

**Comment by julienlancelot** [comment]
ok, then LGTM.

**Pull request closed without merging**